### PR TITLE
Split thread files into generic and architecture specific

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This PR is only supposed to move some code around. All behaviour is supposed to remain unchanged.

There are some small changes where some functions needed to be implemented in a different.
The commit "core: split core/arch/arm/kernel/thread.c" may deserve an extra look and then in particular the functions:
- thread_init_thread_core_local()
- thread_init_per_cpu()

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
The checkpatch complaints are old and I'm not sure it's worth it trying to fix them.

**Comment by jforissier** [comment]
> The checkpatch complaints are old and I'm not sure it's worth it trying to fix them.

I agree, several could be fixed somewhat easily but the code would be a bit messier;  and one at least (`INIT_CANARY()`) would be hard to rewrite. I don't mind ignoring the warnings.

Overall it is a nice cleanup IMO.

- For commits: 
  "core/include/kernel/asan.h: include <compiler.h>"
  "core: move thread_private.h to an include directory"
  "core: split core/arch/arm/include/kernel/thread.h"
  "core: move thread_defs.h into thread.h"
  "core: split core/include/kernel/thread_private.h"
  `Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: split core/arch/arm/kernel/thread.c":
  `Acked-by: Jerome Forissier <jerome@forissier.org>`

FYI, I have run `xtest` with `CFG_CORE_DEBUG_CHECK_STACKS=y` (QEMU and QEMUv8), no issues found.

**Comment by jenswi-linaro** [comment]
Tags applied, thanks.

**Comment by jforissier** [comment]
@maroueneboubakri since this is supposed to help with the RISC-V port you are working on, please feel free to comment.

**Comment by jenswi-linaro** [comment]
Rebased and fixed some conflict.

**Comment by maroueneboubakri** [comment]
Thank you @jenswi-linaro for this PR, it will help a lot, still one function to handle which is `pgt_init()` at https://github.com/jenswi-linaro/optee_os/blob/thread/core/kernel/thread.c#L453

**Comment by jenswi-linaro** [comment]
> Thank you @jenswi-linaro for this PR, it will help a lot, still one function to handle which is `pgt_init()` at https://github.com/jenswi-linaro/optee_os/blob/thread/core/kernel/thread.c#L453

That function isn't part of thread.c friends, also you moved the pgt_cache.h header file to core/include/mm yourself so I was assuming you had some other plans for the pgt_ functions.

Anyway, it's better to not delay this PR much longer since the changes here are rather prone to conflict and easy to mess up while fixing those.

**Comment by maroueneboubakri** [comment]
The PR works for me @jforissier

**Comment by jforissier** [comment]
@maroueneboubakri alright, thanks, I'll wait a few minutes before merging in case you would like to give an `Acked-by:` for these commits but there is no obligation of course.

**Comment by maroueneboubakri** [comment]
`Acked-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

File: core/arch/arm/include/kernel/thread_private_arch.h [hunk 1]
```diff
@@ -240,15 +183,11 @@ uint32_t thread_get_usr_sp(void);
 #endif /*ARM32*/
 
 /* Checks stack canaries */
```
Comment (etienne-lms) [inline core/arch/arm/include/kernel/thread_private_arch.h:185]: commit "core: split core/include/kernel/thread_private.h":
could you also remove the inline comment?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with 2 minor comments.

File: core/arch/arm/include/kernel/thread_arch.h [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+ * Copyright (c) 2016-2017, Linaro Limited
+ * Copyright (c) 2020-2021, Arm Limited
+ */
+
+#ifndef __KERNEL_THREAD_ARCH_H
+#define __KERNEL_THREAD_ARCH_H
+
+#ifndef __ASSEMBLER__
+#include <arm.h>
+#include <types_ext.h>
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/thread_arch.h:13]: `<types_ext.h>` already includes stdbool.h and that's something we can depend on.
I'll fix the ordering.

File: core/arch/arm/include/kernel/thread_private_arch.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/thread_private_arch.h:185]: Thanks, I'll fix.

**Comment by jenswi-linaro** [comment]
Comments addressed and tags applied.

**Pull request merged**